components/datetime/src/format/mod.rs

```diff
@@ -95,6 +99,17 @@ pub fn write_pattern<T>(
 where
     T: DateTimeType,
 {
+    // A formatted time is only noon- or midnight-compatible if the
+    // most granular time being displayed lands exactly on the hour.
+    // e.g. 12:05:15 is noon-compatible if displaying only the hour,
+    // but not noon-compatible if displaying the minutes or seconds.
+    let noon_midnight_compatible = match pattern.most_granular_time() {
+        None | Some(TimeGranularity::Hours) => true,
+        Some(TimeGranularity::Minutes) => u8::from(date_time.minute()) == 0,
+        Some(TimeGranularity::Seconds) => {
+            u8::from(date_time.minute()) + u8::from(date_time.second()) == 0
+        }
+    };
```

Comment:
Since `b` is a rare field, I'd suggest to calculate all of that lazily only if you encounter `b`. You could have `get_time_granularity_for_pattern(pattern) -> Option<Granularity>` function for that.